frame/elections-phragmen/src/lib.rs

```diff
@@ -346,42 +421,67 @@ decl_module! {
 			origin,
 			votes: Vec<T::AccountId>,
 			#[compact] value: BalanceOf<T>,
-		) {
+		) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
 
+			// votes should not be empty and more than `MAXIMUM_VOTE` in any case.
 			ensure!(votes.len() <= MAXIMUM_VOTE, Error::<T>::MaximumVotesExceeded);
 			ensure!(!votes.is_empty(), Error::<T>::NoVotes);
 
 			let candidates_count = <Candidates<T>>::decode_len().unwrap_or(0);
 			let members_count = <Members<T>>::decode_len().unwrap_or(0);
 			let runners_up_count = <RunnersUp<T>>::decode_len().unwrap_or(0);
 
+			// can never submit a vote of there are no members, and cannot submit more votes than
+			// all potential vote targets.
 			// addition is valid: candidates, members and runners-up will never overlap.
 			let allowed_votes = candidates_count + members_count + runners_up_count;
-
 			ensure!(!allowed_votes.is_zero(), Error::<T>::UnableToVote);
 			ensure!(votes.len() <= allowed_votes, Error::<T>::TooManyVotes);
 
 			ensure!(value > T::Currency::minimum_balance(), Error::<T>::LowBalance);
 
-			// first time voter. Reserve bond.
-			if !Self::is_voter(&who) {
-				T::Currency::reserve(&who, T::VotingBond::get())
+			// Reserve bond.
+			let new_deposit = Self::deposit_of(votes.len());
+			let Voter { votes: old_votes, deposit: old_deposit, .. } = <Voting<T>>::get(&who);
+			let maybe_refund = if old_votes.is_empty() {
+				// First time voter. Get full deposit.
+				T::Currency::reserve(&who, new_deposit)
 					.map_err(|_| Error::<T>::UnableToPayBond)?;
-			}
+				None
+			} else {
+				// Old voter.
+				match new_deposit.cmp(&old_deposit) {
+					Ordering::Greater => {
+						// Must reserve a bit more.
+						let to_reserve = new_deposit - old_deposit;
+						T::Currency::reserve(&who, to_reserve)
+							.map_err(|_| Error::<T>::UnableToPayBond)?;
+					},
+					Ordering::Equal => {},
+					Ordering::Less => {
+						// Must unreserve a bit.
+						let to_unreserve = old_deposit - new_deposit;
+						let _remainder = T::Currency::unreserve(&who, to_unreserve);
+						debug_assert!(_remainder.is_zero());
+					},
+				};
+				Some(T::WeightInfo::vote_update(votes.len() as u32))
```

Comment:
I think the benchmark only benchmarks for when new_deposit == old_deposit, but here it is used in any condition

Reply:
Great point, this needs a small fix.

Reply:
addressed, working now on updating the weight values.